workspace/src/localization/localization_py/localization_py/EKF.py

```diff
@@ -0,0 +1,134 @@
+        self.R = np.diag([R[0], R[0], R[1]]) ** 2
+        self.P = np.eye(4)
+
+
+    def angle_diff(self,ref, act):
+        """Computes the difference between two angles.
+
+        Computes the difference between two angles in radians using modular subtraction. This solves the issue when we have a positive and negative angle, both with very large magnitude.
+
+        Args:
+            ref: The reference angle in radians.
+            act: the actual angle in radians.
+        
+        Returns:
+            The difference between the two angles, using modular arithmatic. For example, an input of {pi, pi/2} will return pi/2. An input of {-pi+epsilon, pi-epsilon} will return 2 epsilon.
+        """
+        
+        if( (ref>0 and act>0) or (ref<=0 and act <=0)):
+            err_theta = ref-act
+        elif( ref<=0 and act > 0):
+            if(abs(ref-act)<abs(2*np.pi+ref-act)):
+                err_theta = -abs(act-ref)
+            else:
+                err_theta = abs(2*np.pi + ref- act)
+        else:
+            if(abs(ref-act)<abs(2*np.pi-ref+act)):
+                err_theta = abs(act-ref)
+            else:
+                err_theta = -abs(2*np.pi-ref+act)
+        return err_theta
+
+    def predict(self, x, u):
+        """The preidction phase of the EKF.
+
+        Updates the state based off of the motion model, and ensures that angles are in the range [-pi, pi]. Updates the linearized motion model, and the predicted covariance matrix.
+
+        Args:
+            x: The 4-D current state vector of the vehicle.
+            u: The 2-D input vector for the vehicle.
+
+        Returns:
+            The 4-D updated state of the vehicle based off of the motion model.
+        """
+
+        step_size = 1
+        self.dt = self.dt/step_size
+        for i in range(0,step_size):
+            x = self.dyn.motion_model(x, u)
+
+        if(x[2,0]>np.pi):
+            x[2,0] = x[2,0]-2*np.pi
+        if(x[2,0]<-np.pi):
+            x[2,0] = x[2,0]+2*np.pi
+        self.dt = self.dt*step_size
+        F = self.dyn.calc_F(x[3][0], x[2][0], u[1][0])
+
+        self.P = F @ self.P @ F.T + self.Q
+        return x
+
+    def correct(self, x, z):
+        """The correction phase of the EKF.
+
+        The observation correction phase for the EKF. The observation model is defined, and then the standard EKF motion model is applied, with angle regularization into the [-pi, pi] range applied intermitently.
+
+        Args:
+            x: The 4-D current state vector of the vehicle.
+            z: The 3-D observation vector of the vehicle, consisting of cartesian coordinates x, y from the GPS measurement, and heading theta from the magnetometer.
+
+        Returns:
+            The corrected state of the vehicle based off of the sensor measurements.
+        """
+        H = np.array([
+            [1, 0, 0, 0],
+            [0, 1, 0, 0],
+            [0, 0, 1, 0]
+        ])
+        zPred = H@x
+        while(z[2,0]>np.pi):
```

Comment:
Why is this a while?

Reply:
Also can you explain to me why you need to do this transform? Is it because the magnetometer only gives values between -pi and pi? If that is the case, then I think its fine to do the transforms in the state estimation function. 

But if its a dynamic model thing, then they should be transformed in the dynamic model class because not all dynamic models will need this.

Reply:
Generally shouldn't be a while, it should only run once if necessary... this is an artifact that I left in from a while ago where if the car spun around multiple times you got measurements that were 4pi+ radians, so to normalize you needed a while loop. This felt safer (in case this problem ever came back due to other circumstances; maybe there are instances where you want to keep track of how many rotations you have done). If you feel strongly about this it can probably be changed to an if statement...